ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+dApps.
+
+However, the ASA lacks in flexibility and configurability. For instance, once
+issued it can't be re-configured (its unit name, decimals, maximum supply).
+Also, it is freely transferable (unless frozen). This prevents developers from
+specifying additional business logic to be checked while transferring it (think
+of royalties or [vesting](https://en.wikipedia.org/wiki/Vesting)).
+
+Enforcing transfer conditions require freezing the asset and transferring it
+through a clawback operation -- which results in a process that is opaque to
+users and wallets and a bad experience for the users.
+
+The Smart ASA defined by this ARC extends the ASA to increase its expressiveness
+and its flexibility. Developers can now introduce additional business logic
+around its operations. Wallets, dApps, and SDKs can recognize Smart ASAs and
+adjust their flows and user experiences accordingly.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC
+2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+The following sections describe:
+
+- The ABI interface for a controlling Smart Contract (the Smart Contract that
+  controls a Smart ASA).
+- The metadata required to denote a Smart ASA and define the association between
+  an ASA and its controlling Smart Contract.
+
+### ABI Interface
+
+The ABI interface specified here draws inspiration from the transaction
+[reference](https://developer.algorand.org/docs/get-details/asa/#asset-functions)
+of an Algorand Standard Asset (ASA).
+
+To provide a unified and familiar interface between the Algorand Standard Asset
+and the Smart ASA, method names and parameters have been adapted to the ABI
+types but left otherwise unchanged.
+
+#### Asset Creation
+
+```json
+{
+  "name": "asset_create",
+  "args": [
+    { "type": "uint64", "name": "total" },
+    { "type": "uint32", "name": "decimals" },
+    { "type": "bool", "name": "default_frozen" },
+    { "type": "string", "name": "unit_name" },
+    { "type": "string", "name": "name" },
+    { "type": "string", "name": "url" },
+    { "type": "byte[]", "name": "metadata_hash" },
+    { "type": "address", "name": "manager_addr" },
+    { "type": "address", "name": "reserve_addr" },
+    { "type": "address", "name": "freeze_addr" },
+    { "type": "address", "name": "clawback_addr" }
+  ],
+  "returns": { "type": "uint64" }
+}
+```
+
+Calling `asset_create` creates a new Smart ASA and returns the identifier of the
+ASA. The [metadata section](#metadata) describes its required properties.
+
+> Upon a call to `asset_create`, a reference implementation SHOULD:
+>
+> - Mint an Algorand Standard Asset (ASA) that MUST specify the properties
+>   defined in the [metadata section](#metadata). In addition:
+>   - The `manager`, `reserve` and `freeze` addresses SHOULD be set to the
```

Comment:
This would preclude using ARC-19 in conjunction with ARC-20, but maybe that's ok since ARC-20 allows you to have essentially a dynamic URL anyway? In which case, it raises the question of what `SHOULD` the smart contract add for the URL field in the ASA itself, and furthermore, if you wanted to support a metadata standard with traits for a smart ASA which standard(s) do we want to support? Should the ABI for `asset_create` and `asset_config` support taking transaction note, or should we say that the smart ASA tracked URL field should always be used (forcing you to store metadata off-chain e.g. IPFS). When box storage comes along that maybe gives a third option that's completely on-chain?

If we do say that the URL field should always be looked up via the smart contract rather than the ASA then possibly we can repurpose the URL field on the ASA to indicate it's a smart ASA (e.g. `algorand://arc-20` or `algorand://{app id of managing contract}#arc-20` or something), rather than the unit name / name as indicated below (which I personally think would be a bit messy given it's the default bit of text you would see about any ASA while viewing it and waiting for other data to load about the ASA (e.g. while looking up IPFS / indexer etc.).

Reply:
Hmm, just read the "Specifying the controlling Smart Contract" section. Given this contract controls the creation of the smart contract my point about supporting `note` stands since if you want to use ARC-69 (unsure if the contract trusts that the identification of the smart contract is included in the ARC-69 payload or not (I guess it has no choice but to trust it for ARC-3 anyway).

Reply:
>This would preclude using ARC-19 in conjunction with ARC-20, but maybe that's ok since ARC-20 allows you to have essentially a dynamic URL anyway?

@robdmoore ARC-20 is so general that you can dynamically change whatever `field` you want, not just the `URL`, also the `name`, the `unit_name` and so on.

The binding between the "underlying ASA" and the Smart ASA App can be done in several ways, personally I'm support the option to use an AVM-accessible field to programmatically verify the binding, for example: the `URL` field of the "underlying ASA", not to be confused with the `URL` of the Smart ASA, which could be set to whatever value you want (even bigger than 96 kB) and changed over time.

With respect to traits: in addition to regular Smart ASA field you may add to the Smart ASA state whatever traits you want and even have a specific on-chain logic to change each trait, which is accessible to the AVM (just like any variable).

Reply:
Fair enough, and that would make sense to be a separate ARC standard in the future so marketplaces have a standard way of determining traits.

Reply:
(but to enable ARC-69 support this interface needs transaction note I think).

Reply:
The ARC-20 does not prevent addition of a `note` along with the `InnerTxn` resulting from its ABI calls. So it's maybe more a matter of the ARC-69 to specify that in case the asset is created as Smart ASA the implementation must take care of assembling the `note` of the `InnerTxn` according to the ABI call that generates the `AssetCreate` or `AssetConfig`. WDYT?